core/src/packet_forward_manager.rs

```diff
@@ -0,0 +1,427 @@
+use {
+    crate::{
+        banking_stage::TOTAL_BUFFERED_PACKETS, qos_service::QosService,
+        unprocessed_packet_batches::*,
+    },
+    solana_perf::packet::Packet,
+    solana_runtime::{bank::Bank, cost_model::TransactionCost, cost_tracker::CostTracker},
+    solana_sdk::transaction::SanitizedTransaction,
+    std::sync::Arc,
+};
+
+#[derive(Debug)]
+pub struct PacketForwardManager<'a> {
+    cost_trackers: Vec<CostTracker>,
+    forwardable_packets: Vec<Vec<&'a Packet>>,
+}
+
+impl<'a> PacketForwardManager<'a> {
+    pub fn new(max_forward_block_count: u64) -> Self {
+        Self {
+            cost_trackers: (0..max_forward_block_count)
+                .into_iter()
+                .map(|_| CostTracker::default())
+                .collect(),
+            forwardable_packets: (0..max_forward_block_count)
+                .into_iter()
+                .map(|_| Vec::<&'a Packet>::new())
+                .collect(),
+        }
+    }
+
+    pub fn take(
+        &mut self,
+        unprocessed_packet_batches: &'a UnprocessedPacketBatches,
+        working_bank: Arc<Bank>,
+        qos_service: &'a QosService,
+    ) -> Vec<&'a Packet> {
+        let prioritized_forwardable_packet_locators = Self::prioritize_unforwarded_packets_by_fee(
+            unprocessed_packet_batches,
+            Some(working_bank.clone()),
+        );
+
+        // if we have a bank to work with, then sort packets by write-account buckets
+        let (transactions, sanitized_locators) = sanitize_transactions(
+            unprocessed_packet_batches,
+            &prioritized_forwardable_packet_locators,
+            &working_bank.feature_set,
+            working_bank.vote_only_bank(),
+            working_bank.as_ref(),
+        );
+        let transactions_costs = qos_service.compute_transaction_costs(transactions.iter());
+
+        transactions
+            .iter()
+            .zip(transactions_costs.iter())
+            .zip(sanitized_locators.iter())
+            .for_each(|((tx, cost), packet_locator)| {
+                if let Some(packet) =
+                    Self::locate_packet(unprocessed_packet_batches, packet_locator)
+                {
+                    self.sort_into_buckets(tx, cost, packet)
+                }
+            });
+
+        let mut result = Vec::<&'a Packet>::new();
+        self.forwardable_packets
+            .iter()
+            .for_each(|v| result.extend(v));
+        result
```

Comment:
this would probably be more efficient as a `iter().collect::<Vec::<&'a Packet>()>()`` instead of multiple calls to `extend()`